ExtiPin trait for GPIOs

Windfisch opened the pull request:
This addresses #35 by porting the ExtiPin trait from the https://github.com/stm32-rs/stm32f4xx-hal repo. (It already contains my recent fix that was discussed over there yesterday.)

Before merging, I would be very happy about a short review :).
If you'd like me to squash the commits together, please tell me.

burrbull commented:
LGTM. But it will be even better if you add example in `example` dir.

Windfisch commented on src/gpio.rs:
```diff
@@ -188,6 +210,69 @@ macro_rules! gpio {
+                                w.bits((r.bits() & !(0xf << offset)) | ($extigpionr << offset))
+                            });
+                        },
+                        _ => {}
```
Good point, thanks. Since this was copypasted from stm32f4xx-hal, it should probably changed there and in stm32h7x-hal.

Windfisch commented:
> LGTM. But it will be even better if you add example in `example` dir.

Good point, I'll add one.

geomatsi commented on src/gpio.rs:
```diff
@@ -518,6 +538,51 @@ macro_rules! gpio {
+                    fn trigger_on_edge(&mut self, exti: &mut EXTI, edge: Edge) {
+                        match edge {
+                            Edge::RISING => {
+                                exti.rtsr.modify(|r, w| unsafe { w.bits(r.bits() | (1 << $i)) });
```
There already exist access methods for all the registers. For instance, example for EXTI1 w/o macros woud look like this: 
```
        exti.imr.modify(|_, w| w.mr1().set_bit());
        exti.ftsr.modify(|_, w| w.tr1().set_bit());
        exti.rtsr.modify(|_, w| w.tr1().set_bit());
```
Is it possible to use the same access methods in order to avoid the use of ```unsafe``` block ?

Windfisch replied:
That's a good point, I was thinking about that myself.
However, I don't think using the mr1/mr2/mr... registers is feasible, since that would make it very difficult to generate such code in the first place. Currently, the macro for the pins just gets a pin number and with that, it is able to set the proper bits for imr, ftsr and rtsr (and other stuff outside of my patch).
Doing anything else would make the macro invocations much more lengthy and, more importantly, error-prone. It would be easy to slip in a copy-paste-mistake where pin3 uses mr3, but tr2 by accident.

Also, since this is inside of a library, I don't see it as a problem here to use `unsafe`. All in all I'd say it's not worth the hassle.

TheZoq2 replied:
Sounds reasonable to me

geomatsi replied:
Sounds reasonable to me as well. Thanks for clarification!

geomatsi commented on src/gpio.rs:
```diff
@@ -269,7 +269,7 @@ macro_rules! gpio {
                 /// Clear the interrupt pending bit for this pin
                 fn clear_interrupt_pending_bit(&mut self, exti: &EXTI) {
-                    exti.pr.modify(|r, w| unsafe { w.bits(r.bits() | (1 << self.i)) });
+                    exti.pr.modify(|_r, w| unsafe { w.bits(1 << self.i) });
```
Maybe just ```_``` instead of ```_r```: 
```
exti.pr.modify(|_, w| unsafe { w.bits(1 << self.i) });
```

Windfisch replied:
that part should have been replaced by `.write` already.

geomatsi commented on src/gpio.rs:
```diff
@@ -518,6 +538,51 @@ macro_rules! gpio {
+                        afio.$exticri.modify(|r, w| unsafe {
+                            let mut exticr = r.bits();
+                            exticr = (exticr & !(0xf << offset)) | ($extigpionr << offset);
+                            w.bits(exticr)
```
There already exist access methods for ```AFIO_EXTICR``` registers. For instance, example for EXTI10_15 w/o macros woud look like this:
```
        // select PD15 as source input for line EXTI10_15
        afio.exticr4
            .exticr4()
            .modify(|_, w| unsafe { w.exti15().bits(0b0011) });
```
So is it possible to simplify suggested implementation a bit using ```exti``` accessors ?

Windfisch replied:
The same as for the `imr, rtsr, ftsr` registers applies here: It would be possible to use these accessors, but at the cost of having to repeat the pin number a couple of times in each of the n*16 pin definitions (n being the number of ports), as `imrN`, `rtsrN`, `exticrM` and so on.
I don't think it would simplify things.

geomatsi commented:
I would also suggest to add an example. Otherwise LGTM. Thanks !

Windfisch commented:
I've added an example. Additionally I implemented a `check_interrupt()` function that checks whether this EXTI pin has actually caused an interrupt. While this is obvious for EXTI1..4, this check is required for EXTI5-9 (which all share one ISR handler).

I'm open for suggestions about the `check_interrupt()` function, the name doesn't yet feel right.

TheZoq2 commented on examples/exti.rs:
```diff
@@ -0,0 +1,68 @@
+    let p = pac::Peripherals::take().unwrap();
+    let cp = cortex_m::peripheral::Peripherals::take().unwrap();
+    {
+        // the scope ensures that the exti reference is dropped before the first ISR can be executed.
```
I'm assuming you do this to avoid a race condition when between `main` and the interrupt handler.

However, I don't think *holding* a reference to something can trigger a race condition, that can only happen if you actually use the reference.

Or is this more of a warning to users that they shouldn't use exti after nvic.enable?

Windfisch replied:
It' both. It's a warning, and also: Isn't it UB that two mutable references to the same data *exist*? (Regardless of them being used?)

TheZoq2 replied:
I'm not entirely sure. My guess is that it's not for the same reason that working with pointers is safe until you actually dereference them, which requires `unsafe` but I may be wrong

TeXitoi replied:
With non lexical lifetime, it's safe if you don't use the mutable reference before enabling interrupt, as it would be accepted by the compiler to create another `&mut` after the enable line.

For examples, I prefer a static mut option to a static mut maybeuninit, less unsafe things to think about.

TeXitoi commented on examples/exti.rs:
```diff
@@ -0,0 +1,68 @@
+        let mut afio = p.AFIO.constrain(&mut rcc.apb2);
+
+        let led = unsafe { &mut *LED.as_mut_ptr()};
+        *led = gpioc.pc13.into_push_pull_output(&mut gpioc.crh);
```
```rust
let led = gpioc.pc13.into_push_pull_output(&mut gpioc.crh);
unsafe { LED = MaybeUninit::new(led) };
```

TeXitoi replied:
no need of scope with this.

TeXitoi commented on examples/exti.rs:
```diff
@@ -0,0 +1,68 @@
+        *exti = gpioa.pa7.into_floating_input(&mut gpioa.crl);
+        exti.make_interrupt_source(&mut afio);
+        exti.trigger_on_edge(&p.EXTI, Edge::RISING_FALLING);
+        exti.enable_interrupt(&p.EXTI);
```
Same pattern:
```rust
let exti = gpioa.pa7.into_floating_input(&mut gpioa.crl);
exti.make_interrupt_source(&mut afio);
exti.trigger_on_edge(&p.EXTI, Edge::RISING_FALLING);
exti.enable_interrupt(&p.EXTI);
unsafe { EXTI = MaybeUninit::new(exti) };
```

TheZoq2 commented:
@Windfisch This is pretty close to getting merged, would you mind fixing the new conflicts.

Also, a changelog entry would be nice :)

Windfisch commented:
thanks for the feedback, @TheZoq2. I'll have a look at it

Windfisch commented:
@TheZoq2, I've resolved the merge conflict and added a Changelog entry.

Do you prefer me to squash together all my changes, or merge it as is? (Or can you just check a "squash" checkbox in Github?)

TheZoq2 commented:
Thanks!

There is a nice big "Squash and merge" button that I can use :)

TheZoq2 commented:
@Windfisch Is there a reason you used a SCREAMING_SNAKE_CASE name for the `Rising` enum? The compiler now gives a warning for that, so if there isn't a particular reason for it, I would like to change it

Windfisch commented:
Hi @TheZoq2, I took most of the code from the stm32f4xx-hal implementation, and they use SCREAMING_SNAKE_CASE, too: 
https://github.com/stm32-rs/stm32f4xx-hal/blob/e94c88ec85488445bbef10542e21173a99781364/src/gpio.rs#L77
For compatibility reasons, I suggest to keep it the same in all hals.

TheZoq2 commented:
Ah alright, i'll add a #[allow(...)] to that enum then :)
